The number of handle columns exceeds the length of `columns_info`

## General Question
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 51.0 failed 4 times, most recent failure: Lost task 0.3 in stage 51.0 (TID 2116) : com.pingcap.tikv.exception.TiClientInternalException: Error reading region:
	at com.pingcap.tikv.operation.iterator.DAGIterator.doReadNextRegionChunks(DAGIterator.java:190)
	at com.pingcap.tikv.operation.iterator.DAGIterator.readNextRegionChunks(DAGIterator.java:167)
	at com.pingcap.tikv.operation.iterator.DAGIterator.hasNext(DAGIterator.java:113)
	at org.apache.spark.sql.tispark.TiRowRDD$$anon$1.hasNext(TiRowRDD.scala:70)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage2.columnartorow_nextBatch_0$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage2.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:755)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.shuffle.sort.UnsafeShuffleWriter.write(UnsafeShuffleWriter.java:177)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: com.pingcap.tikv.exception.RegionTaskException: Handle region task failed:
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at com.pingcap.tikv.operation.iterator.DAGIterator.doReadNextRegionChunks(DAGIterator.java:185)
	... 19 more
Caused by: com.pingcap.tikv.exception.RegionTaskException: Handle region task failed:
	at com.pingcap.tikv.operation.iterator.DAGIterator.process(DAGIterator.java:233)
	at com.pingcap.tikv.operation.iterator.DAGIterator.lambda$submitTasks$1(DAGIterator.java:91)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more
Caused by: com.pingcap.tikv.exception.GrpcException: [components/tidb_query_executors/src/index_scan_executor.rs:109]: The number of handle columns exceeds the length of `columns_info`
	at com.pingcap.tikv.region.RegionStoreClient.handleCopResponse(RegionStoreClient.java:734)
	at com.pingcap.tikv.region.RegionStoreClient.coprocess(RegionStoreClient.java:681)
	at com.pingcap.tikv.operation.iterator.DAGIterator.process(DAGIterator.java:220)
	... 7 more

--- a/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
+++ b/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
@@ -28,7 +28,10 @@ class IssueTestSuite extends BaseTiSparkTest {
     tidbStmt.execute(
       s"CREATE TABLE $dbTable (`CI_NO` varchar(64) NOT NULL, `AC_DT` bigint(20) NOT NULL, `YM_DT` bigint(20) NOT NULL,`SYS_TYPE` varchar(20) NOT NULL, PRIMARY KEY (`CI_NO`,`AC_DT`,`YM_DT`,`SYS_TYPE`) /*T![clustered_index] CLUSTERED */, KEY `IDX_FLOW_01` (`CI_NO`,`AC_DT`))")
     tidbStmt.execute(s"insert into $dbTable values('1',1,1,'1')")
-    spark.sql(s"select ci_no,ac_dt from $dbTable").show()
+    spark.sql(s"select ci_no,ac_dt from $dbTable").explain()
+    spark.sqlContext.setConf(TiConfigConst.USE_INDEX_SCAN_FIRST, "true")
+    spark.sql(s"select * from $dbTable").show()
+    spark.sqlContext.setConf(TiConfigConst.USE_INDEX_SCAN_FIRST, "false")
   }
 
   test("test like escape") {

--- a/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
+++ b/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
@@ -22,6 +22,14 @@ import org.apache.spark.sql.functions.{col, sum}
 
 class IssueTestSuite extends BaseTiSparkTest {
 
+  test("test column mismatch, issue 2750") {
+    val dbTable = "tispark_test.column_mismatch"
+    tidbStmt.execute(s"drop table if exists $dbTable")
+    tidbStmt.execute(
+      s"CREATE TABLE $dbTable (`CI_NO` varchar(64) NOT NULL, `AC_DT` bigint(20) NOT NULL, `SRC_KEY` varchar(100) NOT NULL, PRIMARY KEY (`SRC_KEY`,`CI_NO`,`AC_DT`) /*T![clustered_index] CLUSTERED */, KEY `IDX_FLOW_01` (`CI_NO`,`AC_DT`))")
+    spark.sql(s"select ci_no,ac_dt from $dbTable ").show()
+  }
+
   test("test like escape") {
     val dbTable = "tispark_test.like_escape"
     tidbStmt.execute(s"drop table if exists $dbTable")

--- a/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
+++ b/core/src/test/scala/org/apache/spark/sql/IssueTestSuite.scala
@@ -26,8 +26,9 @@ class IssueTestSuite extends BaseTiSparkTest {
     val dbTable = "tispark_test.column_mismatch"
     tidbStmt.execute(s"drop table if exists $dbTable")
     tidbStmt.execute(
-      s"CREATE TABLE $dbTable (`CI_NO` varchar(64) NOT NULL, `AC_DT` bigint(20) NOT NULL, `SRC_KEY` varchar(100) NOT NULL, PRIMARY KEY (`SRC_KEY`,`CI_NO`,`AC_DT`) /*T![clustered_index] CLUSTERED */, KEY `IDX_FLOW_01` (`CI_NO`,`AC_DT`))")
-    spark.sql(s"select ci_no,ac_dt from $dbTable ").show()
+      s"CREATE TABLE $dbTable (`CI_NO` varchar(64) NOT NULL, `AC_DT` bigint(20) NOT NULL, `YM_DT` bigint(20) NOT NULL,`SYS_TYPE` varchar(20) NOT NULL, PRIMARY KEY (`CI_NO`,`AC_DT`,`YM_DT`,`SYS_TYPE`) /*T![clustered_index] CLUSTERED */, KEY `IDX_FLOW_01` (`CI_NO`,`AC_DT`))")
+    tidbStmt.execute(s"insert into $dbTable values('1',1,1,'1')")
+    spark.sql(s"select ci_no,ac_dt from $dbTable").show()
   }
 
   test("test like escape") {

--- a/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
@@ -227,7 +227,14 @@ public class TiDAGRequest implements Serializable {
     indexScanBuilder.setTableId(getPhysicalId()).setIndexId(indexInfo.getId());
     if (tableInfo.isCommonHandle()) {
       for (TiIndexColumn col : tableInfo.getPrimaryKey().getIndexColumns()) {
-        indexScanBuilder.addPrimaryColumnIds(tableInfo.getColumn(col.getName()).getId());
+        long primaryColumnId = tableInfo.getColumn(col.getName()).getId();
+        indexScanBuilder.addPrimaryColumnIds(primaryColumnId);
+        // add primary columns to columns:
+        // https://github.com/pingcap/tidb/blob/ca7ba14da87abd060922b32c893b121c4c3d6373/pkg/executor/builder.go#L569
+        if (!isPrimaryInColumns(indexScanBuilder.getColumnsList(), primaryColumnId)) {
+          TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
+          indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
+        }
       }
     }
 
@@ -236,6 +243,16 @@ public class TiDAGRequest implements Serializable {
         dagRequestBuilder, isDoubleRead, outputOffsets, colOffsetInFieldMap);
 
     return buildRequest(dagRequestBuilder, outputOffsets);
+  }
+
+  private boolean isPrimaryInColumns(
+      java.util.List<com.pingcap.tidb.tipb.ColumnInfo> columnsList, long primaryColumnId) {
+    for (com.pingcap.tidb.tipb.ColumnInfo column : columnsList) {
+      if (column.getColumnId() == primaryColumnId) {
+        return true;
+      }
+    }
+    return false;
   }
 
   private void addIndexLookUpIndexRangeScanExecutorCols(

--- a/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
@@ -227,14 +227,11 @@ public class TiDAGRequest implements Serializable {
     indexScanBuilder.setTableId(getPhysicalId()).setIndexId(indexInfo.getId());
     if (tableInfo.isCommonHandle()) {
       for (TiIndexColumn col : tableInfo.getPrimaryKey().getIndexColumns()) {
-        long primaryColumnId = tableInfo.getColumn(col.getName()).getId();
-        indexScanBuilder.addPrimaryColumnIds(primaryColumnId);
+        TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
+        indexScanBuilder.addPrimaryColumnIds(columnInfo.getId());
         // add primary columns to columns:
-        // https://github.com/pingcap/tidb/blob/ca7ba14da87abd060922b32c893b121c4c3d6373/pkg/executor/builder.go#L569
-        if (!isPrimaryInColumns(indexScanBuilder.getColumnsList(), primaryColumnId)) {
-          TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
-          indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
-        }
+        // https://github.com/pingcap/tidb/blob/ddcaadbb856f0890e91e4c77991f0d2aa5aa93d0/pkg/planner/core/planbuilder.go#L1515
+        indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
       }
     }
 
@@ -243,16 +240,6 @@ public class TiDAGRequest implements Serializable {
         dagRequestBuilder, isDoubleRead, outputOffsets, colOffsetInFieldMap);
 
     return buildRequest(dagRequestBuilder, outputOffsets);
-  }
-
-  private boolean isPrimaryInColumns(
-      java.util.List<com.pingcap.tidb.tipb.ColumnInfo> columnsList, long primaryColumnId) {
-    for (com.pingcap.tidb.tipb.ColumnInfo column : columnsList) {
-      if (column.getColumnId() == primaryColumnId) {
-        return true;
-      }
-    }
-    return false;
   }
 
   private void addIndexLookUpIndexRangeScanExecutorCols(

--- a/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
@@ -227,8 +227,7 @@ public class TiDAGRequest implements Serializable {
     indexScanBuilder.setTableId(getPhysicalId()).setIndexId(indexInfo.getId());
     if (tableInfo.isCommonHandle()) {
       for (TiIndexColumn col : tableInfo.getPrimaryKey().getIndexColumns()) {
-        TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
-        indexScanBuilder.addPrimaryColumnIds(columnInfo.getId());
+        indexScanBuilder.addPrimaryColumnIds(tableInfo.getColumn(col.getName()).getId());
       }
     }
 
@@ -306,9 +305,7 @@ public class TiDAGRequest implements Serializable {
         TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
         // add primary columns to columns:
         // https://github.com/pingcap/tidb/blob/ddcaadbb856f0890e91e4c77991f0d2aa5aa93d0/pkg/planner/core/planbuilder.go#L1515
-        if (!isDoubleRead()) {
-          indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
-        }
+        indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
       }
     }
   }

--- a/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java
@@ -229,9 +229,6 @@ public class TiDAGRequest implements Serializable {
       for (TiIndexColumn col : tableInfo.getPrimaryKey().getIndexColumns()) {
         TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
         indexScanBuilder.addPrimaryColumnIds(columnInfo.getId());
-        // add primary columns to columns:
-        // https://github.com/pingcap/tidb/blob/ddcaadbb856f0890e91e4c77991f0d2aa5aa93d0/pkg/planner/core/planbuilder.go#L1515
-        indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
       }
     }
 
@@ -302,6 +299,16 @@ public class TiDAGRequest implements Serializable {
         throw new DAGRequestException(
             "columns other than primary key and index key exist in fields while index single read: "
                 + columnInfo.getName());
+      }
+    }
+    if (tableInfo.isCommonHandle()) {
+      for (TiIndexColumn col : tableInfo.getPrimaryKey().getIndexColumns()) {
+        TiColumnInfo columnInfo = tableInfo.getColumn(col.getName());
+        // add primary columns to columns:
+        // https://github.com/pingcap/tidb/blob/ddcaadbb856f0890e91e4c77991f0d2aa5aa93d0/pkg/planner/core/planbuilder.go#L1515
+        if (!isDoubleRead()) {
+          indexScanBuilder.addColumns(ColumnInfo.newBuilder(columnInfo.toProto(tableInfo)));
+        }
       }
     }
   }

